### PR TITLE
Feature: "Named Transform Image" Component

### PR DESCRIPTION
**Pull request by dherges**
The screenshot should clarify the title:

screen shot 2014-11-02 at 19 53 59

Goal: authors are able to select a named transform config in an image component's edit dialog. Then, the image is delivered by the Named Transform Image Servlet.

In the Touch UI, we can leverage the datasource feature to get the list of available transforms and show them in the authoring dialog. Starting from there, it should be easy to build more complex components for project-specific requirements. I am thinking of features like a HTML5 Picture Component, letting authors choose a transform for different screen resolutions, i.e. edit dialog for  [``srcset`` and ``media`` syntax][1].

I've scrambled together a quick proof-of-concept (see below).

Any ideas or feedback on that? If you like it, I'll write up the code and open a pull request.

``/apps/wip/components/image-transformed/_cq_dialog.xml``
```xml
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:cq="http://www.day.com/jcr/cq/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0" xmlns:nt="http://www.jcp.org/jcr/nt/1.0"
    jcr:primaryType="nt:unstructured"
    jcr:title="Image"
    sling:resourceType="cq/gui/components/authoring/dialog">
    <content
        jcr:primaryType="nt:unstructured"
        sling:resourceType="granite/ui/components/foundation/container">
        <layout
            jcr:primaryType="nt:unstructured"
            sling:resourceType="granite/ui/components/foundation/layouts/tabs"
            type="nav"/>
        <items jcr:primaryType="nt:unstructured">

        	<!-- .. original tabs from wcm/foundation/components/image .. -->

            <appearance
                jcr:primaryType="nt:unstructured"
                jcr:title="Appearance"
                sling:resourceType="granite/ui/components/foundation/section">
                <layout
                    jcr:primaryType="nt:unstructured"
                    sling:resourceType="granite/ui/components/foundation/layouts/fixedcolumns"/>
                <items jcr:primaryType="nt:unstructured">
                    <column
                        jcr:primaryType="nt:unstructured"
                        sling:resourceType="granite/ui/components/foundation/container">
                        <items jcr:primaryType="nt:unstructured">
                            <transform
                                jcr:primaryType="nt:unstructured"
                                sling:resourceType="granite/ui/components/foundation/form/select"
                                fieldDescription="Auslieferungsgröße des Bilds (Breite in Pixel)"
                                fieldLabel="Named Image Transform"
                                name="./transform">
                                <datasource
                                    jcr:primaryType="nt:unstructured"
                                    sling:resourceType="wip/components/image-transformed/datasource"/>
                            </transform>
                        </items>
                    </column>
                </items>
            </appearance>
        </items>
    </content>
</jcr:root>
```

``apps/wip/components/datasource/datasource.esp``
```javascript
<%
/* global sling */
var SimpleDataSource = Packages.com.adobe.granite.ui.components.ds.SimpleDataSource
  , ArrayList = Packages.java.util.ArrayList
  , HashMap = Packages.java.util.HashMap
  , ValueMapResource = Packages.com.adobe.granite.ui.components.ds.ValueMapResource
  , ResourceMetadata = Packages.org.apache.sling.api.resource.ResourceMetadata
  , ValueMapDecorator = Packages.org.apache.sling.api.wrappers.ValueMapDecorator


var configAdmin = sling.getService(Packages.org.osgi.service.cm.ConfigurationAdmin)


var filterName = 'service.factoryPid'
var filterValue = 'com.adobe.acs.commons.images.impl.NamedImageTransformerImpl'
var configs = configAdmin.listConfigurations('(' + filterName + '=' + filterValue + ')')


var resourceList = new ArrayList()
for (var i = 0, len = configs.length; i < len; i++) {
  var transformName = configs[i].getProperties().get('name')

  var map = {
    'value': transformName,
    'text': transformName
  }

  var valMapResource = new ValueMapResource(resource.getResourceResolver(), new ResourceMetadata(), "", new ValueMapDecorator(map));
  resourceList.add(valMapResource);
}

request.setAttribute('com.adobe.granite.ui.components.ds.DataSource', new SimpleDataSource(resourceList.iterator()));

%>
```

``/apps/wip/components/image-transformed/image-transformed.html``
```html
<img src="${resource.path}.transform/${properties.transform}/image.png" >
```

[1]: http://scottjehl.github.io/picturefill/#media-and-srcset "picturefill"

### CODE REVIEW

**Comment by dherges** [comment]
Any chance to give Travis another go at it? error was a "Connection timed out"

**Comment by davidjgonzalez** [comment]
@justinedelson  FYI - I have some minor changes (adding last modified timestamp) to this; am trying to get find time to test it out and get the PR out to @dherges. Ideally would like to get this into next release.

**Comment by dherges** [comment]
@davidjgonzalez perfect, added you to collaborators of dherges/acs-aem-commons - so you should be able to push directly to that repo

**Comment by davidjgonzalez** [comment]
@dherges i made a PR to your fork w a few updates. If/when you merge it, it should flow through to this. Feel free to make changed/adjustments/etc.

**Comment by dherges** [comment]
Looks good to me. Regarding the url pattern: what about to have a service that takes care of generating/validating url patterns for the servlet. So we can take the ".transform/{name}/{stuff}" from the jsp and move it to that service.

See the above commit.

**Comment by davidjgonzalez** [comment]
@dherges yeh - i like it; the thing you have to be careful of is you can point at a bunch of different resource types.

It might make sense to refactor `resolveImage(..)` [1] out as well; Images are a little odd in there are the 2 sides; Writing the Src into the HTML, and then handling the actual image request.

I think this could be a separate PR though.

[1] https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/src/main/java/com/adobe/acs/commons/images/impl/NamedTransformImageServlet.java#L296

**Comment by dherges** [comment]
Ok, I can revert the previous commit and then let's open up another PR. Should I?

**Comment by davidjgonzalez** [comment]
@dherges yeh - might be worth doing. Then we can create new Util (not sure if need a service?) that resolves the Image obj.

TBH i think having a `ImageUtil.resolveImage(resource)` or `ImageUtil.resolveImage(slingRequest)` would be pretty handy outside of Named Image Transform.

File: bundle/src/main/java/com/adobe/acs/commons/wcm/datasources/impl/DataSourceBuilderImpl.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import org.apache.sling.commons.json.JSONObject;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+@Component(
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/wcm/datasources/impl/DataSourceBuilderImpl.java]: @justinedelson any thoughts around this? Seems like creating DataSources is a common pattern and everything is common except for the list of Title/Text pairs.

File: bundle/pom.xml [hunk 2]
```diff
@@ -453,6 +453,11 @@
             <version>2.3.4</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
```
Comment (justinedelson) [inline bundle/pom.xml]: Same comment as in #445 - are we ready to remove 5.6.1 compatibility?

**Comment by davidjgonzalez** [comment]
Add Sightly example to documentation for this component 

```
<!--/**
 * Example: if you are using Sightly components, the component could look something like this
 * the src url is constructed using '${resource.path}.transform/${properties.transform}/image.png'
 */-->
<a
    data-sly-use.image="image.js"
    data-sly-unwrap="${!image.linkUrl}"
    href="${image.linkUrl}"
    onclick="CQ_Analytics.record({event: 'imageClick', values: { imageLink: '${image.linkurl @ context='scriptString'}', imageAsset: '${image.filereference @ context='scriptString'}', imageTitle: '${image.fileName @ context='scriptString'}' }, collect: false, options: { obj: this }, componentPath: '${image.resource.resourceType @ context='scriptString'}'})">
  <img
      src="${resource.path}.transform/${properties.transform}/image.png"
      alt="${image.alt}"
      title="${image.title}"
      width="${image.width}"
      height="${image.height}"
      class="${image.cssClass}"
      data-emptytext="Image"
      usemap="${image.imageMaps.hash}"/>
</a>
<small data-sly-test="${properties.jcr:description}">${properties.jcr:description}</small>
<map
    data-sly-list="${image.imageMaps.maps}"
    name="${image.imageMaps.name}">
  <area
      shape="${item.type}"
      coords="${item.coords}"
      href="${item.href}"
      target="${item.target}"
      alt="${item.alt}"
      title="${item.title}"/>
</map>
```

**Comment by davidjgonzalez** [comment]
Moved merge to PR #456

**Pull request closed without merging**